src/Microsoft.Azure.WebJobs.Host/Bindings/AttributeCloner.cs

```diff
@@ -94,39 +96,48 @@ private BindingDataResolver GetResolver(PropertyInfo propInfo, INameResolver nam
         {
             // Do the attribute lookups once upfront, and then cache them (via func closures) for subsequent runtime usage. 
             object originalValue = propInfo.GetValue(_source);
+            ConnectionStringAttribute connStrAttr = propInfo.GetCustomAttribute<ConnectionStringAttribute>();
             AppSettingAttribute appSettingAttr = propInfo.GetCustomAttribute<AppSettingAttribute>();
             AutoResolveAttribute autoResolveAttr = propInfo.GetCustomAttribute<AutoResolveAttribute>();
             Validator validator = GetValidatorFunc(propInfo, appSettingAttr != null);
 
-            if (appSettingAttr == null && autoResolveAttr == null)
+            if (appSettingAttr == null && autoResolveAttr == null && connStrAttr == null)
             {
                 validator(originalValue);
 
                 // No special attributes, treat as literal. 
                 return (newAttr, bindingData) => originalValue;
             }
 
-            if (appSettingAttr != null && autoResolveAttr != null)
+            int attrCount = new Attribute[] { connStrAttr, appSettingAttr, autoResolveAttr }.Count(a => a != null);
+            if (attrCount > 1)
             {
-                throw new InvalidOperationException($"Property '{propInfo.Name}' cannot be annotated with both AppSetting and AutoResolve.");
+                throw new InvalidOperationException($"Property '{propInfo.Name}' can only be annotated with one of the types {nameof(AppSettingAttribute)}, {nameof(AutoResolveAttribute)}, and {nameof(ConnectionStringAttribute)}.");
             }
 
             // attributes only work on string properties. 
             if (propInfo.PropertyType != typeof(string))
             {
-                throw new InvalidOperationException($"AutoResolve or AppSetting property '{propInfo.Name}' must be of type string.");
+                throw new InvalidOperationException($"{nameof(ConnectionStringAttribute)}, {nameof(AutoResolveAttribute)}, or {nameof(AppSettingAttribute)} property '{propInfo.Name}' must be of type string.");
```

Comment:
Eventually, it would be nice to have an analyzer for this restriction to make sure we're failing at compile time.